### DISCUSSION TITLE
Crop Tool and Aspect Ratio Bucketing

### DISCUSSION THREAD

**Discussion by HuskedHull** [FennelFetish/qapyq#17] [General]
Scale and Crop need to do something about bucketing.

**Comment by FennelFetish** [comment]
Can you elaborate?

The size presets for the Crop Tool can be changed in the Crop tab in the Batch Window.
This also changes the presets for the Scale Tool but these are not applied until a restart.
I'll add something so the size presets can be changed directly in the tools without the batch window.

**Comment by HuskedHull** [comment]
I do not really understand bucketing.
SD 1.5 guides on CivitAI (most popular guides mention SDXL, but use other models) say that images will be cropped and resized, and Kohya GUI has a random crop option.
I guess an image's resolution, original or after resizing, must fit a bucket exactly, or it will be put into the biggest bucket it does NOT fit [EDIT: the smallest bucket it fits or barely fits with one side?], and then cropped. And a bucket's side lengths are defined in steps of 64 or so.
So if I have a 4000x3000 image, it has an aspect ratio of 1.333, and pixel number 11.444091797 bigger than 1 mebipixel (1024x1024).
There is no bucket with the aspect ratio 1.333.
1152÷896 = 1.286
1216÷832 = 1.46154

3000÷(4000÷1216) = 912
4000÷(3000÷832) = 1109.333333333
3000÷(4000÷1152) = 864
4000÷(3000÷896) = 1194.666666667
4000÷(3000÷768) = 1024
3000÷(4000÷1344) = 1008

My configuration (I tried omitting the default and the resolutions I added trying to solve this issue)
```
        "1024x1024",
        "1088x960",
        "1152x896",
        "1216x832",
        "1344x768",
        "1472x704",
        "1600x640",
        "1792x576",
        "2048x512",
        "960x1088",
        "896x1152",
        "832x1216",
        "768x1344",
        "704x1472",
        "640x1600",
        "576x1792",
        "512x2048",
```

If these are correct, it can be resized to
1344x1008 - 1354752 has to be cropped to fit a bucket
1195x896 - 1070720 same
1109x912 - 1011408 - has to be cropped despite not having too many pixels.
1152x864 - fits in 1152x896, but has to be either augmented to be used in it, or cropped to fit into a smaller bucket
1024x768 - fits in many buckets, but none perfectly

How can qapyq help me decide what to do?

**Comment by HuskedHull** [comment]
I found the Quantized mode, but why would I want to deform the image?
It doesn't show/compare the pixel count.
Other modes don't show the aspect ratio.

**Comment by FennelFetish** [comment]
I also have a lot of open questions regarding bucketing.

What I've gathered:
- The number of images in a size bucket needs to be divisible by the batch size. If it is not, OneTrainer will drop images. Kohya will repeat images.
- Some guides for these tools say that images don't need to be cropped or resized at all. Automatic bucketing should handle this. I'd like to know how it does that. I also read that a large number of different sizes and different aspect ratios can lead to problems, such as underfilled buckets. That would make sense

I *think* the images are rescaled to match `resolution^2` total pixels while keeping the aspect ratio. And then the images are bucketed together based only on their aspect ratio, which then are also the same size.
And the size is probably quantized to match the VAE's constraints?

I'm not sure the random cropping is for making images fit. OneTrainer also has that, and it's for variation/augmentation. There's a paper that states training with random crops, and also multiple resolutions, improves quality. The random crops only use a small offset, so it doesn't cut away large parts of the training data.


I've just had another atempt at reading the OneTrainer code for this.
This is weird: https://github.com/Nerogar/mgds/blob/d74523c250393dc79f7384403f5f88bd5b13269b/src/mgds/pipelineModules/AspectBucketing.py#L74
I'll investigate further.

I crop most images to 1024x1536 or swapped. The main reason is that I could reuse the dataset in the future for models that are based on larger resolutions. And if needed, I could resize and auto-crop these images to smaller sizes and other aspect ratios - or let the training tool do that.
I don't only crop to match the buckets. I also crop away unwanted things, select a composition, and I use the crop tool for selecting the images that I want to include in the dataset.

But I'm quite sure that we shouldn't be quantizing the images manually.

The other scaling modes don't show the aspect ratio because they keep it.

**Comment by HuskedHull** [comment]
> I'm not sure the random cropping is for making images fit.

https://github.com/bmaltais/kohya_ss/blob/master/docs/LoRA/options.md#dont-upscale-bucket-resolution
https://github.com/bmaltais/kohya_ss/blob/master/docs/LoRA/options.md#random-crop-instead-of-center-crop

**Comment by HuskedHull** [comment]
> I don't only crop to match the buckets. I also crop away unwanted things, select a composition,

I forgot to mention that I want only a part of the 4000x3000 image, so I need to crop it even if it fits an aspect ratio, and fit a bucket and possibly choose a size that would scale down well.

**Comment by FennelFetish** [comment]
There are other things to consider when choosing dimensions:

It is said that the models generate the best images when using the dimensions that it was trained on. Image dimensions certainly has a big effect on composition, but I never noticed a difference in quality. The only thing I noticed in this regard is that close-ups of faces become uncanny when using a too large resolution (in SD1.5 and SDXL). This matters for upscaling and inpainting.
I think this is related to the weaker self-attention in models without transformers, which also leads to repeated elements, such as limbs and heads, when generating the base image with too large resolutions.

There are other workarounds and model patches for this, but could composition be fixed with a LoRA that includes high-resolution images?
And regarding image quality, I always upscale images so I'd want to optimize for large resolutions.

Resolution affects training speed.

I guess it could make sense to crop to a *multiple* of the model's favorite dimensions. Like when it favors 892x1152, we could crop large images to twice this size (1784x2304) and smaller images to 892x1152. This for building the dataset, but for training the large images are scaled by a factor of 0.5, which means less quality loss, and we'd keep the option to reuse the images for models with larger dimensions. 

And when the image dimensions are prepared manually beforehand, and the buckets are manually filled, it could make sense to disable the automatic bucketing and hence further scaling and cropping (if the size is properly quantized and if not doing multi-resolution training).

The selected resolution in the training tool is relevant and we'd have to consider how the images scale down and how they must be cropped for quantization (resolution steps).

Another option that avoids further cropping is to pad the images to a different aspect ratio and use a mask to exclude the padding. This has to consider the 8x8 compression factor of VAEs and latents, meaning the mask has to be quantized and this might still exclude small parts of the image.

Or we just rely on the automatic cropping of the training tools, keep it in mind and account for this using a large enough border when cropping manually.

I also found this: (https://github.com/kohya-ss/sd-scripts/blob/main/docs/train_SDXL-en.md)
> `--bucket_reso_steps` can be set to 32 instead of the default value 64. Smaller values than 32 will not work for SDXL training.

**Comment by HuskedHull** [comment]
Those old guides recommend cropping high-resolution images to a portait or a "cowboy shot" to improve faces. Such images supposedly make the LoRA generate portraits and "cowboy shots" more often, which seems like the reason they suggest those kinds of close-ups, but do they affect face quality in *full body shots*?

What if I crop to legs? It will generate legs, but will that be of any use to me?
Should I tag a few images both with "person, scene" and "person's head, person's arms, person's torso, person's legs, [scene if not masked out]", and the leg image with "person's legs"? What size should the legs on the legs-only image be?

Similarly, if there are pictures of multiple smaller subjects, will closer pictures of each of the subjects shown separately help those?

**Comment by HuskedHull** [comment]
quoting only the relevant parts of the comments; the mentions of accumulation steps are only included for context (they are discussing fine-tuning, and there are pros and cons):

https://github.com/Nerogar/OneTrainer/discussions/251#discussioncomment-9088548

>>> [Calamdor](https://github.com/Calamdor) on Apr 12, 2024
>>>
>>> You can try a batch of 4 and accumulation of steps of 4, which gives you a virtual batch of 16, and see how run goes. Batch 4 appears to be the sweet spot, if you can do it, at least for consumer hardware.
>>> 
>>> With batches, you need to watch out for aspect ratios being dropped during bucketing. If you do not have enough images in a bucket, they will not be used. So with 22 images, you can duplicate a few to get 24 and have 6 steps per epoch (assuming all images can be used), or you can do a repeat of 4, and have 22 steps per epoch (ensuring all images will be used).
>>> 
>>> Accumulation steps do not need to be evenly divisible by your steps in an epoch in OneTrainer, which can cause samples and saves to spill into the next epoch if it is not evenly divisible.

>> [Z1-Gamer](https://github.com/Z1-Gamer) on Apr 12, 2024
>>
>>  Do the number of images for each of the resolutions need to be divisible by the batch size

https://github.com/Nerogar/OneTrainer/discussions/251#discussioncomment-9095084

> [Calamdor](https://github.com/Calamdor) on Apr 12, 2024
> 
> Generally speaking. Resolutions, no. Aspect ratios, yes. So if you have three 3x4 images at different resolutions, they should be bucketed together at the resolution you are training at (normally 1024px for SDXL). That still would not fill a batch size of 4 though. This is where the repeats can come in to help minimize the work required. Currently, I just pad all my images square to not have to focus so much on this part of prepping the dataset, as I found cropping exhausting after creating a few datasets.
> 
> There are exceptions to the above, multi-res training is one. Using resolution override is another.

**Comment by FennelFetish** [comment]
> Those old guides recommend cropping high-resolution images to a portait or a "cowboy shot" to improve faces. Such images supposedly make the LoRA generate portraits and "cowboy shots" more often, which seems like the reason they suggest those kinds of close-ups, but do they affect face quality in _full body shots_?

From https://github.com/Nerogar/OneTrainer/wiki/Lessons-Learnt-and-Tutorials#dataset-batch-size-and-training-strategy-by-caith

> Caith typically devotes 30% of the dataset to close-up images, 50% to what he wants the final outputs to resemble, and 20% to concepts that might be used alongside his LoRA or fine-tune. This helps the model adapt more effectively.

> - For instance, if the character is associated with gambling, that extra 20% might include images of cards, empty casinos, bustling casinos, bars, and so on.
    When working with a specific style, he adds about 20% of the dataset with images that are not in that style (e.g., cars, people, backgrounds). This acts like targeted “regularization” data rather than random filler.

This was only recently added to the doc. I can confirm that close-ups improve understanding in wide angle view, but it doesn't seem that the model will always make the connection. It might be beneficial to include multiple zooming levels.
And I think some details may be associated with the "close-up of" tokens if used in the caption, and then dropped when not generating close ups. Adding a line in the .txt caption file without "close-up" could possibly help (that's similar to caption dropout).

> What if I crop to legs? It will generate legs, but will that be of any use to me? Should I tag a few images both with "person, scene" and "person's head, person's arms, person's torso, person's legs, [scene if not masked out]", and the leg image with "person's legs"?

Ok I think I misunderstood. Yes that can help with details. Not sure about the captioning method though. As mentioned, I use "close-up of ..." and I think I got this from elsewhere.
This is what I wrote first:

That will teach proportions. The models already have an understanding for terms that describe the composition/perspective, and it's probably better to use these terms. It understands "cowboy shot" and other photography terms (medium shot etc), and the VLMs/LLMs do too, by the way.
When captioning, think about it this way: When I would want to generate this image, what would I want to prompt? If you want to generate portraits by default, you would omit anything about the perspective for portrait images. When including legs or other unwanted or changable elements, you'd tag those. If the model then generates cowboy shots, you can put cowboy shot in the negative prompt. That works.

And then there's also ControlNets for controlling the composition.

> What size should the legs on the legs-only image be?

I try to include either full limbs, or the knees, elbows or other parts with distinct texture. I'm hoping that the model will pick this up as a hint about proportions / orientation / as context.
(I wrote this before I realised the misunderstanding. I'm talking about including the upper body AND knees/feet/arms. But including context certainly makes sense. Think of inpainting - that also needs context.)


> Similarly, if there are pictures of multiple smaller subjects, will closer pictures of each of the subjects shown separately help those?

Probably, but the models are prone to mixing attributes up. I usually inpaint with specialized prompts.


And if you want to train on details, or separate training of details and composition, there's also the option of adjusting noising schedule, timestep distributions, or however it's called, which is another rabbit hole. Don't know if Kohya has settings for that, OneTrainer does and I wish it would allow this as a setting per concept.

**Comment by HuskedHull** [comment]
> OneTrainer does and I wish it would allow this as a setting per concept.

https://github.com/Nerogar/OneTrainer/issues/684

**Comment by FennelFetish** [comment]
I tried OneTrainer without Aspect Ratio Bucketing and that will just center-crop all images to resolution^2, making square images. This can be observed when enabling the debug option.

To effectively disable automatic bucketing and cropping, I used the Stats in qapyq to filter the images by image size. For each size, I open the respective images in a new tab and use Batch File to create symlinks in a new folder named e.g. "896x1152".
Then in OneTrainer, I add all these folders as a separate concept with Resolution Override in the Image Augmentation tab enabled and set to e.g. "896x1152".
Aspect Ratio Bucketing set to ON, but that's ignored since the resolutions are overridden.

**Comment by HuskedHull** [comment]
If [this](https://github.com/Nerogar/OneTrainer/wiki/Lessons-Learnt-and-Tutorials#what-does-this-mean) is true:
> The more distinct concepts you include (as long as they’re consistently tagged in captions), the larger the batch size you may need to keep those concepts from bleeding into each other.

Do images from concepts that need to be distinguished need to have the same AR?

**Comment by FennelFetish** [comment]
It seems the AR or resolution acts like a strong form of conditioning and I think it's better to include a concept using multiple resolutions. Otherwise it might only appear when generating with that specific resolution.
(For composition-related concepts. Pretty sure details depend less on the resolution. Or put differently: High timesteps are affected more by the resolution. I guess!)

The generation resolution has a big effect on composition. I don't think that's directly related to batching.
Choosing distinct resolutions for concepts could be an effective way to force a composition, if that's what you want.

In OneTrainer at least, the concepts are all mixed up and all images of an AR bucket are shuffled to make the batches for training.
Verified by adding one concept with 5 images and another with 3 images, all the same resolution. At batch size 8 an epoch has 1 step, so they're batched together.
It's just a way for adding multiple folders, but the concepts are defined in the captions.

**Comment by HuskedHull** [comment]
Why does "Target Size" (still) select the minimum size of the selection?

Showing the AR near the size would be helpful to choose a similar AR.

**Comment by FennelFetish** [comment]
> Why does "Target Size" (still) select the minimum size of the selection?

I don't understand.

**Comment by HuskedHull** [comment]
> > Why does "Target Size" (still) select the minimum size of the selection?
> 
> I don't understand.

Target Size in Crop,
the image is 720x544
select 1024x1024, selection becomes 544x544
select 512x512, selection stays 544x544, but can be reduced to 512x512 by scrolling.

**Comment by FennelFetish** [comment]
> the image is 720x544
> select 1024x1024, selection becomes 544x544

That's what "Constrain to Image" does. You can disable that and outside pixels will be exported as transparent if the format allows for that.

When target size is 1024x1024 and selection is 544x544, the selection will be scaled to 1024x1024.

**Comment by HuskedHull** [comment]
But what about

> select 512x512, selection stays 544x544, but can be reduced to 512x512 by scrolling.

**Comment by FennelFetish** [comment]
Without "Allow Upscale", upscaling is prevented by the minimum selection size.

**Comment by HuskedHull** [comment]
I see. "Target Size" is the **export** target size, not a selection size.

> Set the target size.
>     Exported images will be exactly this size.

If I change the selection size, the extracted fragment will be rescaled to the target size on export.

The [original issue](https://github.com/FennelFetish/qapyq/issues/10#issuecomment-2674927009) is probably because I didn't notice this, and was also distracted by how that made the orientation toggle work.
Maybe "Target Size" should say "Export Size" instead.
